assignment2/task2.py

```diff
@@ -0,0 +1,84 @@
+# -*- coding: UTF-8 -*-
+
+import math
+
+
+class LCABuilder(object):
+    def __init__(self, graph):
+        self.graph = graph
+        self.root = 0
+        self.two_pow = int(math.ceil(math.log(len(graph), 2)))
+
+        self.time_in = [0] * len(graph)
+        self.time_out = [0] * len(graph)
+        self.parent = [[0] * (self.two_pow + 1) for _ in xrange(len(graph))]
+        self.timer = 1
+
+    def build(self):
+        self._dfs(self.root)
+
+    def _dfs(self, ver, pnt=0):
+        self.time_in[ver] = self.timer
+        self.timer += 1
+
+        self.parent[ver][0] = pnt
+        for index in xrange(1, self.two_pow + 1):
+            prev = self.parent[ver][index - 1]
+            self.parent[ver][index] = self.parent[prev][index - 1]
+        for index in xrange(len(self.graph[ver])):
+            to = self.graph[ver][index]
+            if to == pnt:
+                continue
+            self._dfs(to, ver)
+
+        self.time_out[ver] = self.timer
+        self.timer += 1
+
+    def find(self, first_ver, second_ver):
+        if self._check(first_ver, second_ver):
+            return first_ver
+        if self._check(second_ver, first_ver):
+            return second_ver
+        for step in xrange(self.two_pow, -1, -1):
```

Comment:
If this is not true, starting from power log_2(n) (keep this in `self.two_pow` attribute) try to find highest (closest to root) vertex for A, that not the ancestor of B (that vertex C, for which C is not the ancestor or B, but `self.parent[C][0]` - ancestor of B ). We can find this vertex with O(log n) time complexity.